modules/ruxhal/src/platform/aarch64_common/ivc.rs

```diff
@@ -0,0 +1,218 @@
+ /// if let Err(e) = conn.close() {
+ ///     error!("Error closing connection: {}", e);
+ /// }
+ /// ```
+ 
+ pub fn ivc_example() {
+     let mut conn = Connection::new();
+ 
+     // Establish the connection
+     if let Err(e) = conn.connect(0) {
+         info!("Error connecting: {}", e);
+         return;
+     }
+ 
+     // Send the message
+     if let Err(e) = conn.send_message("Hello from VM!") {
+         info!("Error sending message: {}", e);
+     }
+ 
+     // Close the connection
+     if let Err(e) = conn.close() {
+         info!("Error closing connection: {}", e);
+     }
+ }
+ 
+ pub struct Connection<'a> {
+     // Reference to the allocator
+     allocator: &'a GlobalAllocator, 
+     ivc_info_ptr: *mut IvCInfo,
+     control_table_ptr: *mut ControlTable,
+ }
+ 
+ impl<'a> Connection<'a> {
+     
+     pub fn new() -> Self {
+         // Get the reference to the global allocator
+         let allocator = global_allocator(); 
+         debug!("Connection created.");
+         Connection {
+             allocator,
+             ivc_info_ptr: ptr::null_mut(),
+             control_table_ptr: ptr::null_mut(),
+         }
+     }
+ 
+     pub fn connect(&mut self, communication_zone: usize) -> Result<(), &'static str> {
+         let alloc_size = HVISOR_HC_IVC_INFO_SIZE;
+         let align = HVISOR_HC_IVC_INFO_ALIGN;
+         let layout = Layout::from_size_align(alloc_size, align).unwrap();
+         
+         let ptr = self.allocator.alloc(layout).expect("Memory allocation failed!");
+         self.ivc_info_ptr = ptr.as_ptr() as *mut IvCInfo;
+ 
+         let vpa_ivcinfo = self.ivc_info_ptr as usize;
+         let pa_ivcinfo: usize = vpa_ivcinfo - __PA;
+ 
+         debug!("The memory address of the IVC Info: VA: 0x{:x}, IPA: 0x{:x}", vpa_ivcinfo, pa_ivcinfo);
+ 
+         ivc_hvc_call(HVISOR_HC_IVC_INFO as u32, pa_ivcinfo, HVISOR_HC_IVC_INFO_SIZE);
+         debug!("ivc_hvc_call finished.");
+ 
+         // Safety: At this point we know ivc_info_ptr is valid and allocated
+         let ivc_info: &IvCInfo = unsafe { &*self.ivc_info_ptr };
+ 
+         let control_table_ptr = (ivc_info.ivc_ct_ipas[communication_zone] + __PA as u64) as *mut ControlTable;
+         self.control_table_ptr = control_table_ptr;
+ 
+         info!("IVC Connection established.");
+         Ok(())
+     }
+ 
+     pub fn send_message(&self, message: &str) -> Result<(), &'static str> {
+         if self.ivc_info_ptr.is_null() {
+             return Err("Not connected");
+         }
+     
+         let ivc_info: &IvCInfo = unsafe { &*self.ivc_info_ptr };
+     
+         let control_table: &ControlTable = unsafe { &*self.control_table_ptr };
+     
+         // Suppose we are zone1, we are to send message to zone0.
+         // Therefore use the out_sec_size field of the ControlTable struct (typically 0x1000).
+         let offset = control_table.out_sec_size as u64;
+         let address: u64 = ivc_info.ivc_shmem_ipas[0] + offset + __PA as u64;
```

Comment:
DONT get virtual address via adding __PA, look up at the function of `direct_virt_to_phys`. It might be useful for it.

Reply:
fixed.